### PR TITLE
[mlir][bufferization]-Replace only one use in TensorEmptyElimination

### PR DESCRIPTION
**Pull request by amirBish**
In many cases the emptyTensorElimination can not transform or eliminate the empty tensor which is being inserted into the `SubsetInsertionOpInterface`.

Two major reasons for that:

1- Failing when trying to find a legal/suitable insertion point for the `subsetExtract` which is about to replace the empty tensor. However, we may try to handle this issue by moving the needed values which responsible on building the  `subsetExtract` nearby the empty tensor (which is about to be eliminated). Thus increasing the probability to find a legal insertion point.

2-The EmptyTensorElimination transform replaces the tensor.empty's uses all at once in one apply, rather than replacing only the specific use which was visited in the use-def chain (when traversing from the tensor.insert_slice). This scenario of replacing all the uses of the tensor.empty may lead into additional read effects after bufferization of the specific subset extract/subview which should not be the case.

Both cases may result in many copies in the coming bufferization which can not be canonicalized.

The first case can be noticed when having a `tensor.empty` followed by `SubsetInsertionOpInterface` (or in simple words `tensor.insert_slice`), which have been lowered from `tensor/tosa.concat`.

The second case can be noticed when having a `tensor.empty`, with many uses and leading to applying the transformation only once, since the whole uses have been replaced at once.

The first commit in the PR only adds the lit tests for the cases shown above (NFC), to emphasize how the transform works, in the coming MRs will upload a slight changes to handle these case.

The second commit in this PR, we want to replace only the specific use which was visited in the `use-def` chain (when traversing from the `tensor.insert_slice`'s source).

### CODE REVIEW

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 1]
```diff
@@ -28,47 +28,72 @@ namespace bufferization {
...
+  // Trying to move the needed values before the `emptyTensorOp`.
+  for (Value val : neededValues) {
+    if (valueDominateInsertionPoint(domInfo, candidateInsertionPoint, val))
+      continue;
+    Operation *definingOp = val.getDefiningOp();
+    if (!definingOp)
+      continue;
+
+    bool isItSafeToMoveOp =
+        llvm::all_of(definingOp->getOperands(), [&](Value operand) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:83]: How useful is this actually? There could be situation where a valid insertion point can be found by moving both `definingOp` and the defining ops of `definingOp`'s operands.

I have no better solution but I'm wondering if this is the right approach.

Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:83]: Thanks Matthias for the review, I have started to handle a more trivial case where the needed values can be moved, and hopefully their operands dominate the insertion point or maybe our needed values have no operands and can be moved without any consideration, A useful example for the case i have mentioned is the lowered concatenation.

I think we may start with a simple case and then expand it to handle much more complex issues. let me know what you think.

Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:83]: @matthias-springer Added additional lit test to show this case, where still insertion point may be invalid because of the definingOp of the definingOp's operands.

**Comment by amirBish** [comment]
@matthias-springer ping, could you have additional look on the changes ?

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp (hunk 1 above)
Comment (matthias-springer) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:83]: I'm not really happy with this approach because this is not a general solution. I think this is just solving one edge case. Moving single ops around feels a bit like a hack. What if tomorrow somebody comes along with a use case where we have to move an entire graph of ops (i.e., defining ops of the operands of the op that you're moving)? Are we going to keep extending the pass like that?

But I don't know a better solution that works for all cases.

Can we make this transformation extensible with a lambda, such that downstream users can customize the transformation? Some options that come to mind:
* A lambda that computes the insertion point. By default (for compatibility with existing code), it does what the code is doing now. But users can provide a different lambda that also moves ops around and can thus compute an insertion point in cases such as yours.
* A "buildSubsetExtraction" lambda that internally computes an insertion point, calls `SubsetInsertionOpInterface::buildSubsetExtraction` and `getValuesNeededToBuildSubsetExtraction` and may also move ops around. (This sounds slightly better to me than the previous suggestion.)

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 2]
```diff
@@ -159,8 +194,9 @@ LogicalResult mlir::bufferization::eliminateEmptyTensors(
         replacement = rewriter.create<tensor::CastOp>(v.getLoc(), v.getType(),
                                                       replacement);
       }
-      // Replace the tensor::EmptyOp.
-      rewriter.replaceOp(emptyTensorOp, replacement);
+      // Replace the specific use of the tensor::EmptyOp.
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:197]: This part makes sense to me, this could be a separate PR.

Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:165]: Great, modified the PR to include only this change.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp (hunk 1 above)
Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp]: Nice, For now will only include the changes of the replacing only one use. And after submitting this one will adopt the approach you have suggested to have a lambda of `buildSubsetExtraction` that user can control. I guess it will abstract the different implementation of finding the valid insertion point and let user have custom support.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 3]
```diff
@@ -28,47 +28,72 @@ namespace bufferization {
 using namespace mlir;
 using namespace mlir::bufferization;
 
+/// Return true if `val` is in scope at the given
+/// `insertionPoint`.
+static bool valueDominateInsertionPoint(const DominanceInfo &domInfo,
```
[the remark this answers is not in the document]
Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp]: Yeah, for now after the changes this function does not exist anymore.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 4]
```diff
@@ -48,27 +48,20 @@ neededValuesDominateInsertionPoint(const DominanceInfo &domInfo,
...
-  });
-}
-
-/// Find a valid insertion point for a replacement of `emptyTensorOp`, assuming
-/// that the replacement may use any value from `neededValues`.
+/// Find a valid insertion point for a replacement of `useToBeEliminated`,
+/// assuming that the replacement may use any value from `neededValues`.
 static Operation *
-findValidInsertionPoint(Operation *emptyTensorOp,
+findValidInsertionPoint(OpOperand *useToBeEliminated,
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:54]: From a functional perspective, this is correct, but I think this would be easier to understand if you replace `OpOperand*` with `Operation *emptyTensorOp, Operation *user`.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 5]
```diff
@@ -131,16 +125,25 @@ LogicalResult mlir::bufferization::eliminateEmptyTensors(
...
         source.get(), /*condition=*/
-        [&](Value val) { return val.getDefiningOp<tensor::EmptyOp>(); },
-        config);
+        [&](Value val) { return val.getDefiningOp<tensor::EmptyOp>(); }, config,
+        &visitedOpOperands);
 
     for (Value v : emptyTensors) {
       Operation *emptyTensorOp = v.getDefiningOp();
 
+      // Find the use to be replaced from the use-def chain
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:134]: nit: `.` missing at the end of the sentence.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 6]
```diff
@@ -131,16 +125,25 @@ LogicalResult mlir::bufferization::eliminateEmptyTensors(
...
+      auto iter = llvm::find_if(
+          visitedOpOperands, [&emptyTensorOp](OpOperand *opOperand) {
+            return llvm::count(emptyTensorOp->getUses(), *opOperand);
+          });
+      if (iter == visitedOpOperands.end())
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:139]: Under what circumstances is there no matching op operand? Can you add a comment?

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp [hunk 7]
```diff
@@ -159,8 +162,9 @@ LogicalResult mlir::bufferization::eliminateEmptyTensors(
         replacement = rewriter.create<tensor::CastOp>(v.getLoc(), v.getType(),
                                                       replacement);
       }
-      // Replace the tensor::EmptyOp.
-      rewriter.replaceOp(emptyTensorOp, replacement);
+      // Replace the specific use of the tensor::EmptyOp.
+      useToBeReplaced->getOwner()->setOperand(
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:166]: wrap this in `rewriter.modifyOpInPlace`

**Review by matthias-springer — APPROVED** [review]

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp (hunk 7 above)
Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp]: Great, added.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp (hunk 5 above)
Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp]: Fixed.

File: mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp (hunk 6 above)
Comment (amirBish) [reply mlir/lib/Dialect/Bufferization/Transforms/EmptyTensorElimination.cpp:141]: Just added a comment about it, This could be achieved `emptyTensorOp` is being used by the insertionOp directly.
